Fix reading 0 bits

jmajnert opened the pull request:
I was getting errors where the decoder wanted to decode 0 bits as a field value and fell over. This change adds a graceful failure by emitting and error message

gab-arrobo commented:
ok to test

jmajnert commented:
> I think it would be better to understand where this issue is coming from. Can you please check whether the issue you are seeing if because of lines 505 or 617?

The data was coming from Viavi simulator. I need to set up the environment to be able to reproduce. This may take a while

SeanCondon commented:
> @eroshiva @SeanCondon, what are your thoughts about this comment?

Good point @gab-arrobo - I see a debug statement on line 615, so it should be easy to tell if 617 is the source. @jmajnert  Maybe it is in previously capture logs?

jmajnert commented:
> > @eroshiva @SeanCondon, what are your thoughts about this comment?
> 
> Good point @gab-arrobo - I see a debug statement on line 615, so it should be easy to tell if 617 is the source. @jmajnert Maybe it is in previously capture logs?

I found the logs ([attached here](https://github.com/onosproject/onos-lib-go/files/13809507/integer_crash.txt)). It's line 615

gab-arrobo commented:
> I found the logs ([attached here](https://github.com/onosproject/onos-lib-go/files/13809507/integer_crash.txt)). It's line 615

@jmajnert, can you please update your PR to address the issue before calling line 617?

jmajnert commented:
> @jmajnert, can you please update your PR to accordingly address the issue before calling line 617?

That would mean fixing how `rawLength` is calculated, right?  I'm not an expert on APER. I'd rather not touch this.

gab-arrobo commented:
> That would mean fixing how `rawLength` is calculated, right? I'm not an expert on APER. I'd rather not touch this.

Not necessarily, for example in line [960](https://github.com/onosproject/onos-lib-go/blob/master/pkg/asn1/aper/aper.go#L960), there is an `if statement` to call a function only if its argument is `> 0`. You could do something similar, no?

jmajnert commented:
You're proposing guarding against error conditions in multiple call sites instead of doing it once in the callee.
Why would that be a better solution?

SeanCondon commented:
I think the benefit is that it gives a more precise error message.

gab-arrobo commented:
> You're proposing guarding against error conditions in multiple call sites instead of doing it once in the callee. Why would that be a better solution?

@jmajnert, based on the logs you shared, the issue is because `pd.bytes[pd.byteOffset]` is `0` in line [578](https://github.com/onosproject/onos-lib-go/blob/master/pkg/asn1/aper/aper.go#L578). Is it correct for `pd.bytes[pd.byteOffset]` to be `0`? Should we add a return right after line `578` when `rawLength = 0`?

jmajnert commented:
> @jmajnert, based on the logs you shared, the issue is because `pd.bytes[pd.byteOffset]` is `0` in line [578](https://github.com/onosproject/onos-lib-go/blob/master/pkg/asn1/aper/aper.go#L578). Is it correct for `pd.bytes[pd.byteOffset]` to be `0`? Should we add a return right after line `578` when `rawLength = 0`?

That's exactly what I meant! If the calculation in line 578 is incorrect, it should be fixed there. But I'm not an expert and I don't know how to fix it. If the calculation is correct, meaning that input data is malformed, then an `if` check is enough.
Still I think that `GetBitString` should check for correctness of its arguments.

jmajnert commented:
So far I got these suggestions:
1. Add a `rawLength == 0` check before calling `pd.getBitsValue()`. I added a new PR with this solution here: https://github.com/onosproject/onos-lib-go/pull/247/files
2. "return right after line 578 when rawLength = 0". I added a new PR with this solution here: https://github.com/onosproject/onos-lib-go/pull/248/files

This gives us 3 PRs, they shouldn't be conflicting. Please choose your preferred way of fixing this problem
